src/gatsby-node.ts

```diff
@@ -28,34 +28,62 @@ export async function sourceNodes(
   // get objects
   const s3 = new AWS.S3();
 
-  const listObjects = async bucket => {
-    // todo improve this call
-    // see https://stackoverflow.com/a/49888947
-    const response = await s3
-      .listObjectsV2({
-        Bucket: bucket
-        // todo handle continuation token
-        // ContinuationToken: token,
-      })
-      .promise();
-
-    // add bucket key
-    const objects = response.Contents?.reduce((acc: ObjectType[], cur) => {
-      const object: ObjectType = { ...cur, Bucket: bucket };
-      acc.push(object);
-      return acc;
-    }, []);
-
-    return objects;
+  const getS3ListObjects = async (params: {
+    Bucket: string;
+    ContinuationToken?: string;
+  }) => {
+    return await s3
+      .listObjectsV2(params)
+      .promise()
+      .catch(error => {
+        reporter.error(
+          `Error listing S3 objects on bucket "${params.Bucket}": ${error}`
+        );
+      });
+  };
+
+  const listAllS3Objects = async (bucket: string) => {
+    const allS3Objects: ObjectType[] = [];
+
+    const data = await getS3ListObjects({ Bucket: bucket });
+
+    if (data?.Contents) {
+      data.Contents.forEach(object => {
+        allS3Objects.push({ ...object, Bucket: bucket });
+      });
+    } else {
+      reporter.error(
+        `Error processing objects from bucket "${bucket}". Is it empty?`
+      );
+    }
+
+    let nextToken = data && data.IsTruncated && data.NextContinuationToken;
+
+    while (nextToken) {
+      const data = await getS3ListObjects({
+        Bucket: bucket,
+        ContinuationToken: nextToken
+      });
+
+      if (data && data.Contents) {
+        data.Contents.forEach(object => {
+          allS3Objects.push({ ...object, Bucket: bucket });
+        });
+      }
+      nextToken = data && data.IsTruncated && data.NextContinuationToken;
+    }
+
+    return allS3Objects;
   };
 
   try {
-    let objects: Array<any> = await Promise.all(
-      buckets.map(bucket => listObjects(bucket))
+    const allBucketsObjects: Array<any> = await Promise.all(
```

Comment:
Since:
- `listAllS3Objects` returns an array of `ObjectType`
- Promise.all() puts all `listAllS3Objects` responses in an array

I think we know that `allBucketsObjects` should be an array of arrays of `ObjectType` 🙂 

```suggestion
    const allBucketsObjects: ObjectType[][] = await Promise.all(
```